### PR TITLE
Add Artifactory as release artifact source

### PR DESCRIPTION
**Pull request by barbelity**

### CODE REVIEW

File: vss-extension.json [hunk 1]
```diff
@@ -110,6 +120,125 @@
                 ]
             }
         },
+        {
+            "id": "artifactory-release-artifact-type",
+            "description": "Artifactory",
+            "type": "ms.vss-releaseartifact.release-artifact-type",
+            "targets": [ "ms.vss-releaseartifact.artifact-types" ],
+            "properties": {
+                "name": "Artifactory",
```
Comment (yahavi) [inline vss-extension.json]: Find a way to add Artifactory icon

File: buildScripts/build.js [hunk 2]
```diff
@@ -17,6 +17,17 @@ function installTasks() {
             // If a package.json is missing, npm will exec the install command on the parent folder. This will cause an endless install loop.
             if (fs.existsSync(path.join(taskDir, "package.json"))) {
                 cleanExecNpm('i', taskDir);
+            } else {
+                fs.readdir(taskDir, (err, innerDirs) => {
```
Comment (yahavi) [inline buildScripts/build.js]: Please change innerDirs --> taskVersionsDirs

File: buildScripts/build.js [hunk 3]
```diff
@@ -17,6 +17,17 @@ function installTasks() {
...
+                    if (innerDirs) {
+                        innerDirs.forEach(versToBuild => {
+                            let taskNewDir = path.join(taskDir, versToBuild);
```
Comment (yahavi) [inline buildScripts/build.js]: Please change taskNewDir --> taskVersionDir
or use `taskDir`

File: tasks/ArtifactoryGenericDownload/Ver2/downloadArtifacts.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    } else {
+        console.log("Generic download...");
+        performGenericDownload(cliPath, workDir, artifactoryService, artifactoryUrl);
+    }
+}
+
+function performArtifactSourceDownload(cliPath, workDir, artifactoryService, artifactoryUrl) {
+    let buildNumber = tl.getInput("version", true);
+    let buildName = tl.getInput("definition", true);
+    let downloadPath = tl.getInput("downloadPath", true);
```
Comment (yahavi) [inline tasks/ArtifactoryGenericDownload/Ver2/downloadArtifacts.js]: Please mention in a comment where do you get `downloadPath` from

File: tasks/ArtifactoryGenericDownload/Ver2/task.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        "Patch": "0"
+    },
+    "minimumAgentVersion": "1.83.0",
+    "instanceNameFormat": "Artifactory Generic Download",
+    "inputs": [
+        {
+            "name": "connection",
+            "type": "connectedService:artifactoryService",
+            "label": "Artifactory service",
+            "defaultValue": "",
```
Comment (yahavi) [inline tasks/ArtifactoryGenericDownload/Ver2/task.json]: I think we can omit this field

File: tasks/ArtifactoryGenericDownload/Ver2/downloadArtifacts.js (hunk 4 above)
Comment (yahavi) [reply tasks/ArtifactoryGenericDownload/Ver2/downloadArtifacts.js]: Use `utils.fixWindowsPaths`

File: tasks/ArtifactoryGenericDownload/Ver2/downloadArtifacts.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    buildName = buildName.substr(1);
+
+    let cliCommand = utils.cliJoin(cliPath, cliDownloadCommand, utils.quote("*"), utils.quote(downloadPath), "--build=" + buildName + buildNumber, "--url=" + utils.quote(artifactoryUrl), "--flat=true");
```
[the remark this answers is not in the document]
Comment (barbelity) [reply tasks/ArtifactoryGenericDownload/Ver2/downloadArtifacts.js]: Fixed

File: vss-extension.json [hunk 7]
```diff
@@ -110,6 +120,125 @@
...
+                        ],
+                        "validation": {
+                            "isRequired": true,
+                            "dataType": "string",
+                            "maxLength": 300
+                        }
+                    },
+                    {
+                        "id": "defaultVersionType",
+                        "name": "Build Number",
```
[the remark this answers is not in the document]
Comment (barbelity) [reply vss-extension.json]: Changed this field to look the same as appears in the other artifact sources.

File: buildScripts/build.js (hunk 2 above)
Comment (barbelity) [reply buildScripts/build.js]: Fixed

File: buildScripts/build.js (hunk 3 above)
Comment (barbelity) [reply buildScripts/build.js]: Fixed.

**Pull request merged**